### ISSUE TITLE
Defining a center point for rotating an image

### ISSUE THREAD

**Issue by pearcetm** [openseadragon/openseadragon#2217]
It would be nice to be able to specify the center point/axis around which rotation occurs. For example, when a touch gesture is used to rotate an image, it currently rotates around the center of the viewport even if the gesture is off to one side, which (to me) feels like a non-intuitive user experience. Other use-cases could include a custom control that allows a user to click to define a center point before rotating, so that a feature within the image could be used as the anchor position for the rotation.

I don't see a way to do this easily with the current API, but if I'm missing something I'd be happy to use the existing methods.

I'm not sure what the easiest way to implement this feature would be. Translating the desired center to the viewport center, applying rotation, and translating back to the original coordinates could work. In this case, it would probably be good (at least as an option) to suppress the pan, rotate, and viewport-change events and only issue these events after all the transforms have been applied. I'm open to other ideas too - the simpler the better!

**Comment by iangilman** [comment]
This would be cool! 

We currently have two rotations: One for the viewport and one for each image. I suppose we would want this for both. 

The idea of translating, then rotating, then translating back seems promising; I'd love to know if it works. Are you up for giving it a try?

**Comment by pearcetm** [comment]
It sounds like you're confirming that there's not currently a built-in way to do this :)

I will work on it, likely first as an addon/monkeypatch and see how it works. As I look into whether translate-rotate-untranslate works, is there any way in the existing functionality to suppress events from the viewer during this sequence? If not it may not be a problem, just wondering.

**Comment by iangilman** [comment]
Yeah, there's no built-in way to rotate around a specific point. As far as I know, everything rotates around their center. In the case of the viewport, it's the center of the current view, I believe (rather than the viewport origin). 

We also don't have a way to suppress those events. You could monkey-patch Viewer.raiseEvent and filter them that way. 

Thank you for diving into this!

**Comment by pearcetm** [comment]
Work in progress - this demo shows a tool to rotate around a user-selected point in the viewport. https://pearcetm.github.io/osd-annotation/demo/rotation.html

The translate-rotate-translate back approach is working well. I've taken these steps outside of the viewer code though, and haven't yet figured out how best to update the existing function to add this as an option.

**Comment by iangilman** [comment]
That's looking hot! Awesome demo. Seems nice and smooth. Let me know if you have any questions as you work on integrating it into OSD :)

**Comment by HamzaTatheer** [comment]
Super cool feature!

**Comment by pearcetm** [comment]
@iangilman I starting working on integrating it into OSD but when I synced my branch with the upstream, it pulled in the new changes that enable animating setRotation, and I'm not sure how to merge the two new functionalities.

Here's the implementation that worked well before animation was added:
```

    /**
     * Rotates this viewport to the angle specified.
     * @function
     * @param {Number} degrees The degrees to set the rotation to.
     * @param {?OpenSeadragon.Point} rotationAxisInViewportCoordinates The point to rotate about, in viewport coordinates
     * @return {OpenSeadragon.Viewport} Chainable.
     */
    setRotation: function(degrees, rotationAxisInViewportCoordinates) {
        if (!this.viewer || !this.viewer.drawer.canRotate()) {
            return this;
        }
        let refViewerElementCoordinates;
        if(rotationAxisInViewportCoordinates){
            refViewerElementCoordinates = this.viewportToViewerElementCoordinates(rotationAxisInViewportCoordinates);
            this.panTo(rotationAxisInViewportCoordinates,true);
        }

        this.degrees = $.positiveModulo(degrees, 360);
        this._setContentBounds(
            this.viewer.world.getHomeBounds(),
            this.viewer.world.getContentFactor());
        this.viewer.forceRedraw();

        /**
         * Raised when rotation has been changed.
         *
         * @event rotate
         * @memberof OpenSeadragon.Viewer
         * @type {object}
         * @property {OpenSeadragon.Viewer} eventSource - A reference to the Viewer which raised the event.
         * @property {Number} degrees - The number of degrees the rotation was set to.
         * @property {?Object} userData - Arbitrary subscriber-defined object.
         */
        this.viewer.raiseEvent('rotate', {degrees: degrees});

        if(rotationAxisInViewportCoordinates){
            let refPoint=this.viewerElementToViewportCoordinates(refViewerElementCoordinates);
            let delta = rotationAxisInViewportCoordinates.minus(refPoint);
            this.panBy(delta,true);
        }

        return this;
    },
```

As you can see, there are two conditional steps based on whether an axis of rotation is passed in: panning to the center of rotation, and panning back to the updated center of the viewport after updating this.degrees and calling _setContentBounds.

When this is controlled by a user action, as in my demo, it looks like a smooth rotation even though it was always "immediate". To make this a smooth animation, though, the translate-rotate-translate sequence would have to be followed on every animation update step.

I'm not very familiar with the spring-based animation mechanisms - your input on how this could be done would be much appreciated!

**Comment by iangilman** [comment]
@pearcetm Good point! The actual movement from the springs is handled in the `update` function: 

https://github.com/openseadragon/openseadragon/blob/ed66c2d1518e919dffa335e68c34f469607273b3/src/viewport.js#L1047

You'll have to repeat the rotation axis code in there. You'll need to keep track of the rotation axis in the viewport, since it'll be used beyond just the one `setRotation` call. Take a look at the `zoomPoint` property for some ideas for how to handle that. 

BTW, you might, instead of calling `panTo` and `panBy`, pull the appropriate spring manipulation code out of them (it's just two lines) and that way, you're not sending the events. 

That said, I'm not entirely sure how this is working... Maybe some aspect doesn't translate properly to `update`? Hopefully this at least gets you on the right track...

**Comment by pearcetm** [comment]
@iangilman Thanks for the pointers! I think I found a good way to do this, and created a pull request that adds this functionality. This is currently only for the viewport, not for multi-image. I have not worked with multi-image views, and would be kind of lost trying to figure out what I was doing without a good demo to work with. Presumably the changes would be similar conceptually though. Hopefully if someone who uses that functionality is interested they can take it on.

**Comment by iangilman** [comment]
@pearcetm Yeah, totally fair to leave the multi-image version for the next person :)

**Comment by iangilman** [comment]
Implemented by #2233.

**Comment by grbnzbn** [comment]
@pearcetm Do you know if there is any way to accomplish this outside of the OSD library itself? I'm running v3.1.0 which doesn't have these changes.
